compose/cli/main.py

```diff
@@ -599,38 +599,47 @@ def ps(self, options):
         """
         List containers.
 
-        Usage: ps [options] [SERVICE...]
+        Usage: ps [options] [--filter KEY=VAL...] [SERVICE...]
 
         Options:
-            -q    Only display IDs
+            -q                   Only display IDs
+            --services           Display services
+            --filter KEY=VAL     Filter services by a property (can be used multiple times)
         """
-        containers = sorted(
-            self.project.containers(service_names=options['SERVICE'], stopped=True) +
-            self.project.containers(service_names=options['SERVICE'], one_off=OneOffFilter.only),
-            key=attrgetter('name'))
-
-        if options['-q']:
-            for container in containers:
-                print(container.id)
+        if options['--services']:
+            filters = build_filters(options.get('--filter'))
+            services = self.project.services
+            if filters:
+                services = filter_services(filters, services, self.project)
```

Comment:
Could we filter the containers directly, then iff the `--services` flag is provided, generate a `set()` of services from the filtered container list and print that out? That way, `--filter` could be used with both "modes".

Reply:
This would not work as well. The reason is that `self.project.services` reads the services from the compose file while `self.project.containers` would omit containers that have never been started before.